pkgs/development/tools/godot/default.nix

```
@@ -1,27 +1,28 @@
-{ stdenv, lib, fetchFromGitHub, scons, pkg-config, libX11, libXcursor
-, libXinerama, libXrandr, libXrender, libpulseaudio ? null
+{ stdenv, lib, fetchFromGitHub, scons, pkg-config, libudev ? null, libX11
+, libXcursor , libXinerama, libXrandr, libXrender, libpulseaudio ? null
 , libXi ? null, libXext, libXfixes, freetype, openssl
 , alsaLib, libGLU, zlib, yasm ? null }:
```

```suggestion
{ stdenv, lib, fetchFromGitHub, scons, pkg-config, libudev ? null, libX11
, libXcursor , libXinerama, libXrandr, libXrender, libpulseaudio
, libXi, libXext, libXfixes, freetype, openssl
, alsaLib, libGLU, zlib, yasm }:
```

If those packages are null you are doing something wrong here.

---

Why is that? Am I right in assuming it is because the top level call to `callPackage` should always fill those dependencies in, so they should never be `null`?

In the case of `libudev` and `yasm`, these packages should be optional to the build. Is there a common convention for handling this sort of thing?

---

> Why is that? Am I right in assuming it is because the top level call to callPackage should always fill those dependencies in, so they should never be null?

Yes, they can only be null when you overwrite them to null.

> In the case of libudev and yasm, these packages should be optional to the build. Is there a common convention for handling this sort of thing?

It would be easier to discover if there was a boolean flag like enableUdev or withUdev.

---

When it comes to boolean flags, is there a preference between defaulting to false (opt-in) or defaulting to true (opt-out)? I'd probably go for the former, but is there a preferred approach within nixpkgs?

---

> When it comes to boolean flags, is there a preference between defaulting to false (opt-in) or defaulting to true (opt-out)? I'd probably go for the former, but is there a preferred approach within nixpkgs?

I think we should keep the current behaviour, which should be defaulting to true I believe.

---

Well, I removed the null defaults and updated to 3.3.2, but now the tests seem to be failing. Maybe I'm reading it wrong, but:

`pkgs/development/tools/godot/default.nix:1:1 called without required argument 'libudev'`

seems to indicate that it *is* indeed possible for `libudev` to not exist (making the `null` default necessary). Am I reading this right or is it something else?

---

Then we can default `withUdev` to `false`. From what I see in the latest commit, `libudev` is always put in `buildInputs`, this should be done like `buildInputs = [ ... ] ++ (lib.optional options.withUdev libudev)`, so that it's not needed if `withUdev` is false. Same could be done for other options too.